### DISCUSSION TITLE
Nicely done, thanks for sharing

### DISCUSSION THREAD

**Discussion by neilyoung** [bluenviron/mediamtx#1232] [General]
...going to interface with WebRTC :)

**Comment by aler9** [comment]
WebRTC support is coming in the next days :)

**Comment by neilyoung** [comment]
Oh really? On what basis? I'm going to interface with Kurento Media Server. 

But I would be really keen to see your solution then. Let me know please :)

Kind regards and thanks again. It all literally worked OOB here. I could just not move Raspberry PI (64 bit Bullseye) to consume the RTSP stream.... "invalid frame" with v4l2h264dec and stuff. But don't bother, that's not my use case at all. I just wanted to see the latency

**Reply by aler9** [reply]
https://github.com/aler9/rtsp-simple-server/pull/1242

**Comment by neilyoung** [comment]
I need to shake your hands again: Since Wednesday this week I managed to integrate your solution into a grown WebRTC solution. It now shows up with a big functional enhancement: I'm now able to provide a simple access to a published WebRTC stream via RTSP, RTMP and HLS thanks to your solution. 

I did literally everything with your solution these days:

- Feeding from a spawned FFMPEG session with authentication, connecting it to Kurento Media Server
- Consuming via FFPLAY, Gstreamer, own software clients
- With optional client authentication (using "externalAuthentication" - perfect)
- With optional enrcyption (using Let's Encrypt certificates)

The only open problem is (you know about) the long time missing intra frame in H.264 sessions from browsers.

But anything else worked out of the box. Without any issue. I'm really impressed by the quality of this work. Thanks again for sharing.

BTW: I'm supervising my instances by a NewRelic ping. Would be great, if your code would have some "PING/PONG" entity. Otherwise I can only just check connectivity or look if I find you via "ps -aux".  So this is on my wish list, didn't find it up to now.

Thanks again!

**Reply by aler9** [reply]
Thanks for the appreciation,

> . Would be great, if your code would have some "PING/PONG" entity.

IMHO you can use the metrics API, if you enable metrics and call

```
wget -qO- localhost:9998/metrics
```

The server replies with metrics if it's awake, otherwise the request fails. I use it to implement health checks in Kubernetes.

**Reply by neilyoung** [reply]
Hmm. Tried it right now. Not sure if it works:

```
$ wget localhost:9998/metrics
--2023-01-22 20:04:00--  http://localhost:9998/metrics
Resolving localhost (localhost)... 127.0.0.1
Connecting to localhost (localhost)|127.0.0.1|:9998... failed: Connection refused.

```

**Reply by aler9** [reply]
@neilyoung enable metrics:


```yml
metrics: yes
```

**Reply by neilyoung** [reply]
OK, is there any documentation regarding this? What is happening?

**Reply by aler9** [reply]
https://github.com/aler9/rtsp-simple-server#metrics

**Reply by neilyoung** [reply]
Sorry, found it right now. Maybe this is my missing "accounting interface"

**Reply by neilyoung** [reply]
Out of curiosity: Is this metrics persisted somehow? Does it survive a reboot of the server or the machine?

**Reply by aler9** [reply]
metrics are dynamic in nature, they provides the current state of the server: who is connected, how many bytes have been exchanged and by whom. They're usually fed into a metrics database like Prometheus or InfluxDB, they have the task of providing persistency and can be used to generate graphs and statistics.

**Reply by neilyoung** [reply]
I see. I can see an RTSP session, but I don't see bytecounts.

```
ubuntu@webrtc:~$ cat metrics.4
paths{name="abcde",state="ready"} 1
rtsp_sessions{state="idle"} 0
rtsp_sessions{state="read"} 1
rtsp_sessions{state="publish"} 1
rtsps_sessions{state="idle"} 0
rtsps_sessions{state="read"} 0
rtsps_sessions{state="publish"} 0
rtmp_conns{state="idle"} 0
rtmp_conns{state="read"} 0
rtmp_conns{state="publish"} 0
ubuntu@webrtc:~$ 
```

**Reply by aler9** [reply]
that's because you're using an outdated version, byte counts have been added recently. Upgrade to the latest one.

**Reply by neilyoung** [reply]
Ah ok. Thanks for the hint.
Is this also the case for the http api? Tried this meanwhile.

While this call returns a valid JSON response

```
curl http://127.0.0.1:9997/v1/paths/list
```

this does return 404 (even though I'm running a publisher and a consumer session in parallel

```
ubuntu@webrtc:~$ curl http://localhost:9997/v1/rtspconns/list -v
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9997 (#0)
> GET /v1/rtspconns/list HTTP/1.1
> Host: localhost:9997
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Content-Type: text/plain
< Server: rtsp-simple-server
< Date: Sun, 22 Jan 2023 20:26:07 GMT
< Content-Length: 18
< 
* Connection #0 to host localhost left intact
404 page not foundubuntu@webrtc:~$ 
```

**Reply by neilyoung** [reply]
My version:

wget https://github.com/aler9/rtsp-simple-server/releases/download/v0.20.2/rtsp-simple-server_v0.20.2_linux_amd64.tar.gz

Will update now

**Reply by neilyoung** [reply]
OK. 0.21.1 fixes both problems. 

Question remains: Below the output of the HTTP API (which previously returned 404). I see two sessions, one publisher, one consumer. Is there any way to associate these sessions with paths ("abcde" in my case)?

Output of the metrics request:

```
wget -qO- localhost:9998/metrics
paths{name="abcde",state="ready"} 1
paths_bytes_received{name="abcde",state="ready"} 37313858
rtsp_conns{id="e31b8418-1a7b-4550-bf5e-9213d95e8ce0"} 1
rtsp_conns_bytes_received{id="e31b8418-1a7b-4550-bf5e-9213d95e8ce0"} 1477
rtsp_conns_bytes_sent{id="e31b8418-1a7b-4550-bf5e-9213d95e8ce0"} 533
rtsp_conns{id="7a7eb044-f138-42b2-9c8a-6749586f31d6"} 1
rtsp_conns_bytes_received{id="7a7eb044-f138-42b2-9c8a-6749586f31d6"} 1491
rtsp_conns_bytes_sent{id="7a7eb044-f138-42b2-9c8a-6749586f31d6"} 1599
rtsp_sessions{id="891c4a00-27d8-4ac1-bb05-6ce5b6e19d1f",state="read"} 1
rtsp_sessions_bytes_received{id="891c4a00-27d8-4ac1-bb05-6ce5b6e19d1f",state="read"} 8384
rtsp_sessions_bytes_sent{id="891c4a00-27d8-4ac1-bb05-6ce5b6e19d1f",state="read"} 37063394
rtsp_sessions{id="d6351dfa-3414-41d3-978c-b4e19d2dc29d",state="publish"} 1
rtsp_sessions_bytes_received{id="d6351dfa-3414-41d3-978c-b4e19d2dc29d",state="publish"} 37314586
rtsp_sessions_bytes_sent{id="d6351dfa-3414-41d3-978c-b4e19d2dc29d",state="publish"} 404

```


Output of the HTTP request:

```
{
	"items": {
		"7a7eb044-f138-42b2-9c8a-6749586f31d6": {
			"created": "2023-01-22T20:45:30.350893236Z",
			"remoteAddr": "192.168.188.53:63669",
			"bytesReceived": 2406,
			"bytesSent": 2284
		},
		"e31b8418-1a7b-4550-bf5e-9213d95e8ce0": {
			"created": "2023-01-22T20:45:24.109159524Z",
			"remoteAddr": "127.0.0.1:59762",
			"bytesReceived": 1477,
			"bytesSent": 533
		}
	}
}

```

Can't find it at the first glance in doc

**Reply by neilyoung** [reply]
However, all bytecounts don't make any sense to me... Way to less

**Reply by neilyoung** [reply]
Found an easier way: 

```
curl http://127.0.0.1:8081
```

Returns 404 if the server is up

**Comment by aler9** [comment]
> Is there any way to associate these sessions with paths ("abcde" in my case)?

yes, use the paths/list API:

```
http://localhost:9997/v1/paths/list
```

it returns

```json
{
  "items": {
    "pathname": {
      "confName": "string",
      "source": {
        "type": "rtspSession",
        "id": "string"
      },
      "sourceReady": true,
      "tracks": [
        "string"
      ],
      "bytesReceived": 0,
      "readers": [
        {
          "type": "hlsMuxer"
        }
      ]
    }
  }
}
```

You'll find sessions in the `readers` key.

**Comment by neilyoung** [comment]
Thanks. Appreciated

[a comment by an automated account is not shown]
